### PR TITLE
Two Step Compose Extension and Reactions

### PR DESCRIPTION
**Pull request by vamagra**
@msftdherron @MattSFT @jotrick @goelashish7 @aosolis @jialic 

1. 171875: Two-Step query for compose extension
2. 177737: Add like and reactions into Bot post/message

### CODE REVIEW

File: template-bot-master-csharp/src/controllers/MessagesController.cs [hunk 1]
```diff
@@ -55,14 +54,43 @@ public class MessagesController : ApiController
...
+                var reactionsAdded = activity.ReactionsAdded;
+                var reactionsRemoved = activity.ReactionsRemoved;
+                var replytoId = activity.ReplyToId;
+                Activity reply;
+
+                if (reactionsAdded != null && reactionsAdded.Count > 0)
+                {
+                    reply = activity.CreateReply(Strings.LikeMessage);
+                }
+                else
```
Comment (jotrick) [inline template-bot-master-csharp/src/controllers/MessagesController.cs]: Change this to be the same check but for reactionsRemoved with an else if. Next, move the ReplyToActivityAsync call into the blocks where the reply is set so that, if the reactionsAdded count is 0 and the reactionsRemoved count is 0, then we don't post anything to the channel. (This is in case, in the future, they add another kind of reaction or something)

File: template-bot-master-csharp/src/controllers/MessagesController.cs [hunk 2]
```diff
@@ -55,14 +54,43 @@ public class MessagesController : ApiController
...
             else if (activity.Type == ActivityTypes.Invoke) // Received an invoke
             {
                 // Handle ComposeExtension query
                 if (activity.IsComposeExtensionQuery())
                 {
                     // this will handle the compose extension request
-                    var invokeResponse = WikipediaComposeExtension.GetComposeExtensionResponse(activity);
-                    return Request.CreateResponse<ComposeExtensionResponse>(HttpStatusCode.OK, invokeResponse);
+
+                    if (activity.Name == "composeExtension/selectItem")
```
Comment (jotrick) [inline template-bot-master-csharp/src/controllers/MessagesController.cs]: Add a comment explaining what actually falls into this block.

File: template-bot-master-csharp/utility/TemplateUtility.cs [hunk 3]
```diff
@@ -38,60 +38,69 @@ public static string GetLocale(Activity activity)
...
+        public static ComposeExtensionAttachment CreateComposeExtensionCardsAttachments(WikiHelperSearchResult wikiResult, string state)
         {
-            return GetComposeExtensionMainResultAttachment(title, text, imageUrl, state).ToComposeExtensionAttachment(GetComposeExtensionPreviewAttachment(title, text, imageUrl, state));
+            return GetComposeExtensionMainResultAttachment(wikiResult, state).ToComposeExtensionAttachment(GetComposeExtensionPreviewAttachment(wikiResult, state));
         }
 
-        public static Attachment GetComposeExtensionMainResultAttachment(string title,string text,string imageUrl, string state)
+        public static Attachment GetComposeExtensionMainResultAttachment(WikiHelperSearchResult wikiResult, string state)
         {
+            var tapAction = new CardAction("invoke", value: "{ \"" + "imageUrl" + "\": \"" + wikiResult.imageUrl + "\", \"" + "highlightedTitle" + "\": \"" + wikiResult.highlightedTitle + "\"}");
```
Comment (jotrick) [inline template-bot-master-csharp/utility/TemplateUtility.cs]: I don't believe a tap action needs to be added to the content card in order for the one step multi-turn M.E. to work. Is there a another reason why you add it here?

File: template-bot-master-csharp/utility/TemplateUtility.cs [hunk 4]
```diff
@@ -38,60 +38,69 @@ public static string GetLocale(Activity activity)
...
                     },
                 }.ToAttachment();
             }
         }
 
-        public static Attachment GetComposeExtensionPreviewAttachment(string title, string text, string imageUrl, string state)
+        public static Attachment GetComposeExtensionPreviewAttachment(WikiHelperSearchResult wikiResult, string state)
         {
+            string quoted = cleanForJSON(wikiResult.text);
+            var tapAction = new CardAction("invoke", value: "{ \"" + "imageUrl" + "\": \"" + wikiResult.imageUrl + "\",\"" + "text" + "\": \"" + quoted + "\", \"" + "highlightedTitle" + "\": \"" + wikiResult.highlightedTitle + "\"}");
```
Comment (jotrick) [inline template-bot-master-csharp/utility/TemplateUtility.cs]: Please try to clean up this line. There are inconsistent spaces, unneeded quotes, etc. Would a one line string literal make this cleaner?

File: template-bot-master-csharp/utility/TemplateUtility.cs [hunk 5]
```diff
@@ -114,5 +123,68 @@ public static string ParseInvokeRequestJson(string inputString)
...
+        {
+            if (s == null || s.Length == 0)
+            {
+                return "";
+            }
+
+            char c = '\0';
+            int i;
+            int len = s.Length;
+            StringBuilder sb = new StringBuilder(len + 4);
```
Comment (jotrick) [inline template-bot-master-csharp/utility/TemplateUtility.cs]: Why + 4?

File: template-bot-master-csharp/utility/TemplateUtility.cs [hunk 6]
```diff
@@ -114,5 +123,68 @@ public static string ParseInvokeRequestJson(string inputString)
             return null;
         }
 
+        public static string cleanForJSON(string s)
```
Comment (jotrick) [inline template-bot-master-csharp/utility/TemplateUtility.cs]: Why was this added? Did you encounter issues with the text returned from Wikipedia?

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 7]
```diff
@@ -25,6 +25,8 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
             StateClient stateClient = activity.GetStateClient();
             BotData userData = stateClient.BotState.GetUserData(activity.ChannelId, activity.From.Id);
 
+            var translationHistory = new List<WikiHelperSearchResult>();
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Why do you call this translationHistory?

Comment (jotrick) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Can we just set this below when we get it from the user data?

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 8]
```diff
@@ -115,11 +117,32 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
             {
                 //Signin Experience, please uncomment below code for Signin Experience
-                //composeExtensionResponse = WikiHelper.GetSignin(composeExtensionResponse);
+                //composeExtensionResponse = GetSignin(composeExtensionResponse);
                 //return composeExtensionResponse;
 
                 composeExtensionResponse = new ComposeExtensionResponse();
-                composeExtensionResponse.ComposeExtension = GetMessageResponseResult(Strings.ComposeExtensionInitialRunText);
+
+                var userSession = userData.GetProperty<List<WikiHelperSearchResult>>("ComposeExtensionSelectedResults");
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Can we find a new name for this as well with a comment mentioning what this data is for?

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 9]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+            }
+            else
+            {
+                composeExtensionResponse.ComposeExtension = GetMessageResponseResult(Strings.ComposeExtensionInitialRunText);
+            }
+
+            return composeExtensionResponse;
+        }
+
+        public static WikiHelperSearchResult SetWikiSearchResult(string imageUrl, string highlightedTitle, string cardText)
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Can this just be a constructor for the WikiHelperSearchResult class?

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 10]
```diff
@@ -115,11 +117,32 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+                {
+                    translationHistory = userSession;
+
+                    foreach (var searchResult in translationHistory)
+                    {
+                        WikiHelperSearchResult wikiSearchResult = SetWikiSearchResult(searchResult.imageUrl, searchResult.highlightedTitle, searchResult.text);
+
+                        // create the card itself and the preview card based upon the information
+                        // check user preference for which type of card to create
+                        lstComposeExtensionAttachment.Add(TemplateUtility.CreateComposeExtensionCardsAttachments(wikiSearchResult, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword)));
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Let's get the user's preferred card type above, set it to an easily readable variable like "userPreferredCardType" and pass that into this so it is easier to read. Also make this change to use this variable below.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 11]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+        /// <returns></returns>
+        public static ComposeExtensionResponse HandleComposeExtensionSelectedItem(Activity activity)
+        {
+            WikiHelperSearchResult selectedItem = new WikiHelperSearchResult();
+            var translationHistory = new List<WikiHelperSearchResult>();
+
+            BotData userData = TemplateUtility.GetBotDataObject(activity);
+
+            //Get the Max number of History items from config file
+            int maxTranslationHistory = Convert.ToInt32(ConfigurationManager.AppSettings["MaxTranslationHistoryCount"]);
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: When renaming "translationHistory" change this variable name as well.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 12]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+
+            selectedItem = JsonConvert.DeserializeObject<WikiHelperSearchResult>(activity.Value.ToString());
+
+            var userSession = userData.GetProperty<List<WikiHelperSearchResult>>("ComposeExtensionSelectedResults");
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Rename with whatever is used above.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 13]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+
+            translationHistory = userSession != null ? userSession : translationHistory;
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: If this check is needed, does it need to be used in the above place as well?

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 14]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+
+            if (translationHistory != null && translationHistory.Count > 0)
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Comment to say that you are removing any other occurrences of the current selectedItem so there are not duplicates in the most recently used list

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 15]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+
+
+            ComposeExtensionResponse composeExtensionResponse = new ComposeExtensionResponse();
+            List<ComposeExtensionAttachment> lstComposeExtensionAttachment = new List<ComposeExtensionAttachment>();
+
+            if (selectedItem != null)
+            {
+                // create the card itself and the preview card based upon the information
+                // check user preference for which type of card to create
+                lstComposeExtensionAttachment.Add(TemplateUtility.CreateComposeExtensionCardsAttachments(selectedItem, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword)));
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Same as above - create a variable for userPreferredCardType for readability of this line

File: template-bot-master-csharp/utility/TemplateUtility.cs (hunk 4 above)
Comment (aosolis) [reply template-bot-master-csharp/utility/TemplateUtility.cs]: @vamagra Please introduce a type for the payload (value) and use JsonConvert (JSON.NET) to generate the JSON.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 10 above)
Comment (aosolis) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: "results" or "attachments" -- please don't prefix with "lst"

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 16]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
         }
 
+        /// <summary>
+        /// Purpose of this method is to Keep the history of selected item and return the response Compose Extension
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <returns></returns>
+        public static ComposeExtensionResponse HandleComposeExtensionSelectedItem(Activity activity)
+        {
+            WikiHelperSearchResult selectedItem = new WikiHelperSearchResult();
```
Comment (aosolis) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Move declaration down to line 218 where it's assigned to.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 17]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+            translationHistory.Insert(0, selectedItem);
+
+            //Restrict the transaction History with Max Items.
+            if (translationHistory.Count > maxTranslationHistory)
+            {
+                translationHistory = translationHistory.GetRange(0, maxTranslationHistory);
+            }
+
+            //Save the history Items in user Data
+            userData.SetProperty<List<WikiHelperSearchResult>>("ComposeExtensionSelectedResults", translationHistory);
```
Comment (aosolis) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Define constants for these strings.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 18]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+
+            ComposeExtensionResponse composeExtensionResponse = new ComposeExtensionResponse();
+            List<ComposeExtensionAttachment> lstComposeExtensionAttachment = new List<ComposeExtensionAttachment>();
+
+            if (selectedItem != null)
+            {
+                // create the card itself and the preview card based upon the information
+                // check user preference for which type of card to create
+                lstComposeExtensionAttachment.Add(TemplateUtility.CreateComposeExtensionCardsAttachments(selectedItem, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword)));
+                composeExtensionResponse = GetComposeExtenionQueryResult(composeExtensionResponse, lstComposeExtensionAttachment);
```
Comment (aosolis) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Typo: Extension, not Extenion

File: template-bot-master-csharp/utility/TemplateUtility.cs [hunk 19]
```diff
@@ -38,60 +38,69 @@ public static string GetLocale(Activity activity)
...
 
-        public static ComposeExtensionAttachment CreateComposeExtensionCardsAttachments(string title,string text,string imageUrl, string state)
+        public static ComposeExtensionAttachment CreateComposeExtensionCardsAttachments(WikiHelperSearchResult wikiResult, string state)
         {
-            return GetComposeExtensionMainResultAttachment(title, text, imageUrl, state).ToComposeExtensionAttachment(GetComposeExtensionPreviewAttachment(title, text, imageUrl, state));
+            return GetComposeExtensionMainResultAttachment(wikiResult, state).ToComposeExtensionAttachment(GetComposeExtensionPreviewAttachment(wikiResult, state));
         }
 
-        public static Attachment GetComposeExtensionMainResultAttachment(string title,string text,string imageUrl, string state)
+        public static Attachment GetComposeExtensionMainResultAttachment(WikiHelperSearchResult wikiResult, string state)
```
Comment (aosolis) [inline template-bot-master-csharp/utility/TemplateUtility.cs]: Rename "state", and make it an enum type instead of a string.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 20]
```diff
@@ -156,17 +179,97 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+                lstComposeExtensionAttachment.Add(TemplateUtility.CreateComposeExtensionCardsAttachments(wikiSearchResult, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword)));
             }
 
             composeExtensionResponse = GetComposeExtenionQueryResult(composeExtensionResponse, lstComposeExtensionAttachment);
 
             return composeExtensionResponse;
         }
 
+        /// <summary>
+        /// Purpose of this method is to Keep the history of selected item and return the response Compose Extension
```
Comment (aosolis) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: I'd change this to "Handle the callback received when the user selects an item from the results list"
And at the beginning of the method (line 202) add a comment explaining what you do. "Keep a history of recently-selected items in bot user data. The history will be returned in response to the initialRun query."

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 7 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: I choose the variable name translationHistory as it will hold last 10 selection. for your second comment, I have change the code.

Comment (aosolis) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: I think @jotrick 's point is that this has nothing to do with translations.

Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: got it!!

File: template-bot-master-csharp/src/controllers/MessagesController.cs (hunk 1 above)
Comment (vamagra) [reply template-bot-master-csharp/src/controllers/MessagesController.cs]: Done.

File: template-bot-master-csharp/src/controllers/MessagesController.cs (hunk 2 above)
Comment (vamagra) [reply template-bot-master-csharp/src/controllers/MessagesController.cs]: Done

File: template-bot-master-csharp/utility/TemplateUtility.cs (hunk 3 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/TemplateUtility.cs]: Agree, Removed.

File: template-bot-master-csharp/utility/TemplateUtility.cs (hunk 4 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/TemplateUtility.cs]: Done.

File: template-bot-master-csharp/utility/TemplateUtility.cs (hunk 6 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/TemplateUtility.cs]: yeah there was result rendering issue this is why I used this method to clean the JSON,

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 8 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 9 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 10 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 11 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 12 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 13 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: its not necessary, I have removed it.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 14 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 15 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 10 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 16 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 17 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: please explain it more, its not clear to me.

File: template-bot-master-csharp/utility/TemplateUtility.cs (hunk 19 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/TemplateUtility.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 20 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 18 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/Web.config [hunk 21]
```diff
@@ -13,6 +13,8 @@
   <add key="BaseUri" value="" />
 
   <add key="OFFICE_365_TENANT_FILTER" value="#ANY#" />
+  <add key="MaxTranslationHistoryCount" value="10" />
```
Comment (jotrick) [inline template-bot-master-csharp/Web.config]: Change the name of this value in order to match the other variables - something like "MaxComposeExtensionHistoryCount"

File: template-bot-master-csharp/src/controllers/MessagesController.cs [hunk 22]
```diff
@@ -55,14 +54,42 @@ public class MessagesController : ApiController
...
                 // Handle ComposeExtension query
                 if (activity.IsComposeExtensionQuery())
                 {
-                    // this will handle the compose extension request
-                    var invokeResponse = WikipediaComposeExtension.GetComposeExtensionResponse(activity);
-                    return Request.CreateResponse<ComposeExtensionResponse>(HttpStatusCode.OK, invokeResponse);
+                    // Handle compose extension selected item
+                    if (activity.Name == "composeExtension/selectItem")
+                    {
+                        var selectedItemResponse = WikipediaComposeExtension.HandleComposeExtensionSelectedItem(activity);
```
Comment (jotrick) [inline template-bot-master-csharp/src/controllers/MessagesController.cs]: Add a note inside of the if block above the call to the handler that says, "This handler is used to process the event when a user in Teams selects a result from the compose extension result list"

File: template-bot-master-csharp/src/dialogs/examples/auth/vsts/VSTSHelpers.cs [hunk 23]
```diff
@@ -141,7 +141,7 @@ public static String GenerateAuthorizeUrl(ConversationReference conversationRefe
         }
 
         /// <summary>
-        /// Purpose of this method is to get VSTS Profile
+        /// Get The VSTS Profile
```
Comment (jotrick) [inline template-bot-master-csharp/src/dialogs/examples/auth/vsts/VSTSHelpers.cs]: lower case "the"

File: template-bot-master-csharp/utility/TemplateUtility.cs [hunk 24]
```diff
@@ -2,25 +2,26 @@
...
 using Newtonsoft.Json.Linq;
 using System;
-using System.Collections.Generic;
+using System.Text;
 
 namespace Microsoft.Teams.TemplateBotCSharp.Utility
 {
     /// <summary>
-    /// This method is used to Get the local from incoming acitvity payload
+    /// Get the locale from incoming activity payload and Handle Compose Extension methods
```
Comment (jotrick) [inline template-bot-master-csharp/utility/TemplateUtility.cs]: lowercase "handle compose extension"

File: template-bot-master-csharp/utility/TemplateUtility.cs [hunk 25]
```diff
@@ -38,67 +39,101 @@ public static string GetLocale(Activity activity)
...
+                switch (cardType)
                 {
-                    Title = title,
-                    Images =
-                    {
-                        new CardImage(imageUrl)
-                    },
-                }.ToAttachment();
+                    case CardType.hero:
+                      {
```
Comment (jotrick) [inline template-bot-master-csharp/utility/TemplateUtility.cs]: I don't believe the switch blocks need the { } around them. Thus have the setting of the cardAttachment and the break; statement.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 26]
```diff
@@ -115,11 +121,31 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+                var historySearchWikiResult = userData.GetProperty<List<WikiHelperSearchResult>>("ComposeExtensionSelectedResults");
+                if (historySearchWikiResult != null)
+                {
+                    foreach (var searchResult in historySearchWikiResult)
+                    {
+                        WikiHelperSearchResult wikiSearchResult = new WikiHelperSearchResult(searchResult.imageUrl, searchResult.highlightedTitle, searchResult.text);
+
+                        // create the card itself and the preview card based upon the information
+                        // check user preference for which type of card to create
+                        var userPreferredCardType = TemplateUtility.CreateComposeExtensionCardsAttachments(wikiSearchResult, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword));
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: For this and the same section below:
var userPreferredCardType = userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword));
// optionally if you want to do it in one line or two
var createdCardAttachment = TemplateUtility.CreateComposeExtensionCardsAttachments(wikiSearchResult, userPreferredCardType);
composeExtensionAttachments.Add(createdCardAttachment);

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 27]
```diff
@@ -156,13 +182,83 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
 
                 string cardText = searchResult.snippet + " ...";
 
+                WikiHelperSearchResult wikiSearchResult = new WikiHelperSearchResult(imageUrl, highlightedTitle, cardText);
+
                 // create the card itself and the preview card based upon the information
                 // check user preference for which type of card to create
 
-                lstComposeExtensionAttachment.Add(TemplateUtility.CreateComposeExtensionCardsAttachments(highlightedTitle, cardText, imageUrl, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword)));
+                composeExtensionAttachments.Add(TemplateUtility.CreateComposeExtensionCardsAttachments(wikiSearchResult, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword)));
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Make this match what you make it look like around line 138 - use the variable you set so you only need to get the user's preferred card type once.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 28]
```diff
@@ -156,13 +182,83 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+        /// Keep a history of recently-selected items in bot user data. The history will be returned in response to the initialRun query
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <returns></returns>
+        public static ComposeExtensionResponse HandleComposeExtensionSelectedItem(Activity activity)
+        {
+            BotData userData = TemplateUtility.GetBotDataObject(activity);
+
+            //Get the Max number of History items from config file
+            int MaxHistoryWikiResultCount = Convert.ToInt32(ConfigurationManager.AppSettings["MaxTranslationHistoryCount"]);
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Have this (and other places) use whatever you change this to in the main resource file.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 29]
```diff
@@ -156,13 +182,83 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+
+
+            ComposeExtensionResponse composeExtensionResponse = new ComposeExtensionResponse();
+            List<ComposeExtensionAttachment> composeExtensionAttachment = new List<ComposeExtensionAttachment>();
+
+            if (selectedItem != null)
+            {
+                // create the card itself and the preview card based upon the information
+                // check user preference for which type of card to create
+                var userPreferredCardType = TemplateUtility.CreateComposeExtensionCardsAttachments(selectedItem, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword));
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Handle this the same way as above:
var userPreferredCardType = userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword);
var createdCardAttachment = TemplateUtility.CreateComposeExtensionCardsAttachments(selectedItem, userPreferredCardType);
composeExtensionAttachment.Add(createdCardAttachment);

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 30]
```diff
@@ -156,13 +182,83 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
 
-            composeExtensionResponse = GetComposeExtenionQueryResult(composeExtensionResponse, lstComposeExtensionAttachment);
+            composeExtensionResponse = GetComposeExtensionQueryResult(composeExtensionResponse, composeExtensionAttachments);
+
+            return composeExtensionResponse;
+        }
+
+        /// <summary>
+        /// Handle the callback received when the user selects an item from the results list
+        /// Keep a history of recently-selected items in bot user data. The history will be returned in response to the initialRun query
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Move this comment line ("Keep a history...") into the function itself just above the part of the code where you retrieve the user's most recently used search list so that it explains what that list is for and what you expect to do with it later:

// Keep a history...
BotData userData = TemplateUtility.GetBotDataObject(activity);
var historySearchWikiResult = userData.GetProperty<List<WikiHelperSearchResult>>("ComposeExtensionSelectedResults");
// Get the max number of History items from config file
int maxHistoryWikiResultCount = Convert.ToInt32(Configur.......);

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 28 above)
Comment (jotrick) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: make this variable lower cased at the beginning - maxHistoryWikiResultCount

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 31]
```diff
@@ -156,13 +182,83 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
...
+        public static ComposeExtensionResponse HandleComposeExtensionSelectedItem(Activity activity)
+        {
+            BotData userData = TemplateUtility.GetBotDataObject(activity);
+
+            //Get the Max number of History items from config file
+            int MaxHistoryWikiResultCount = Convert.ToInt32(ConfigurationManager.AppSettings["MaxTranslationHistoryCount"]);
+
+            WikiHelperSearchResult selectedItem = JsonConvert.DeserializeObject<WikiHelperSearchResult>(activity.Value.ToString());
+
+            var historySearchWikiResult = userData.GetProperty<List<WikiHelperSearchResult>>("ComposeExtensionSelectedResults");
```
Comment (jotrick) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Adrian's comment from before was that he wanted "ComposeExtensionSelectedResults" to be set as a constant and then that constant to be used everywhere for this string.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 26 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Can we discuss this comment, didn't understand your intent here?

Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Thanks for the offline explanation, makes perfect sense to me now.

File: template-bot-master-csharp/Web.config (hunk 21 above)
Comment (vamagra) [reply template-bot-master-csharp/Web.config]: Done.

File: template-bot-master-csharp/src/controllers/MessagesController.cs (hunk 22 above)
Comment (vamagra) [reply template-bot-master-csharp/src/controllers/MessagesController.cs]: Done.

File: template-bot-master-csharp/src/dialogs/examples/auth/vsts/VSTSHelpers.cs (hunk 23 above)
Comment (vamagra) [reply template-bot-master-csharp/src/dialogs/examples/auth/vsts/VSTSHelpers.cs]: Done.

File: template-bot-master-csharp/utility/TemplateUtility.cs (hunk 24 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/TemplateUtility.cs]: Done.

File: template-bot-master-csharp/utility/TemplateUtility.cs (hunk 25 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/TemplateUtility.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 26 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 27 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 28 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 29 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 30 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs (hunk 31 above)
Comment (vamagra) [reply template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Done.

File: template-bot-master-csharp/utility/WikipediaComposeExtension.cs [hunk 32]
```diff
@@ -135,8 +137,10 @@ public static ComposeExtensionResponse GetComposeExtensionResponse(Activity acti
 
                         // create the card itself and the preview card based upon the information
                         // check user preference for which type of card to create
-                        var userPreferredCardType = TemplateUtility.CreateComposeExtensionCardsAttachments(wikiSearchResult, userData.GetProperty<string>(Strings.ComposeExtensionCardTypeKeyword));
-                        composeExtensionAttachments.Add(userPreferredCardType);
+
```
Comment (goelashish7) [inline template-bot-master-csharp/utility/WikipediaComposeExtension.cs]: Remove Blank line
